transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,556 @@
+///   machine.
+/// - Transmitting events generated by the [`quinn_proto::Endpoint`] to the corresponding
+///   [`Connection`].
+/// - Receiving messages from the `receiver` and processing the requested actions. This includes
+///   UDP packets to send and events emitted by the [`Connection`] objects.
+/// - Sending new connections on `new_connections`.
+///
+/// When it comes to channels, there exists three main multi-producer-single-consumer channels
+/// in play:
+///
+/// - One channel, represented by `EndpointChannel::to_endpoint` and `receiver`, that communicates
+///   messages from [`EndpointChannel`] to the [`EndpointDriver`].
+/// - One channel per each existing connection that communicates messages from the  [`EndpointDriver`]
+///   to that [`Connection`].
+/// - One channel for the  [`EndpointDriver`] to send newly-opened connections to. The receiving
+///   side is processed by the [`GenTransport`][crate::GenTransport].
+///
+/// In order to avoid an unbounded buffering of events, we prioritize sending data on the UDP
+/// socket over everything else. If the network interface is too busy to process our packets,
+/// everything comes to a freeze (including receiving UDP packets) until it is ready to accept
+/// more.
+///
+/// Apart from freezing when the network interface is too busy, the background task should sleep
+/// as little as possible. It is in particular important for the `receiver` to be drained as
+/// quickly as possible in order to avoid unnecessary back-pressure on the [`Connection`] objects.
+///
+/// ## Back-pressure on `new_connections`
+///
+/// The [`quinn_proto::Endpoint`] object contains an accept buffer, in other words a buffer of the
+/// incoming connections waiting to be accepted. When a new connection is signalled, we send this
+/// new connection on the `new_connection_tx` channel in an asynchronous way, and we only free a slot
+/// in the accept buffer once the element has actually been enqueued on `new_connection_tx`. There
+/// are therefore in total three buffers in play: the `new_connection_tx` channel itself, the queue
+/// of elements being sent on `new_connection_tx`, and the accept buffer of the
+/// [`quinn_proto::Endpoint`].
+///
+/// ## Back-pressure on connections
+///
+/// Because connections are processed by the user at a rate of their choice, we cannot properly
+/// handle the situation where the channel from the background task to individual connections is
+/// full. Sleeping the task while waiting for the connection to be processed by the user could
+/// even lead to a deadlock if this processing is also sleeping waiting for some other action that
+/// itself depends on the background task (e.g. if processing the connection is waiting for a
+/// message arriving on a different connection).
+///
+/// In an ideal world, we would handle a background-task-to-connection channel being full by
+/// dropping UDP packets destined to this connection, as a way to back-pressure the remote.
```

Comment:
> Do you mean by shrinking the MAX_DATA that the remote can send us?

MAX_DATA can't be taken back, once it's issued. It can only ever be increased. Also keep in mind that anything you do takes 1 RTT to take effect.

I don't understand why dropping UDP packets is not safe. In theory (not sure if quinn allows it), you can associate packets with their respective QUIC connection by just parsing the CID and doing a lookup in your connection map. 
I would expect that you don't need any general UDP packet queue here: You read UDP datagrams from the socket, parse their CID (which is _very cheap_), and put them into the queues of their respective QUIC connections. It makes sense that those queues would overflow, if a connection is particularly slow at processing these packets. This would translate into a congestion control signal for the sender, leading to a reduction in the send rate.

Reply:
> > Do you mean by shrinking the MAX_DATA that the remote can send us?
>
> MAX_DATA can't be taken back, once it's issued. It can only ever be increased. Also keep in mind that anything you do takes 1 RTT to take effect.

Ah yes you are right.

---

Discussed out of band with @MarcoPolo and @marten-seemann. As mentioned before we do drop the packets if a channel to a connection is full, which is the best way to solve this issue.